File: RotatedArray.cpp
// Time Complexity : O(logN)
// Space Complexity : O(1) as I didn't use any extra space, just the temporary variables
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Nope
// Problem Link: https://leetcode.com/problems/search-in-rotated-sorted-array/
// Submission Link: https://leetcode.com/submissions/detail/526793424/

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int start = 0, end = nums.size() - 1;
        int mid;
        
        while(start <= end){
            mid = (start + end)/2;
            
            if(target == nums[mid])
                return mid;
            
            // Check if first half of the array is sorted
            if(nums[start] <= nums[mid]){
                // If the target lies in first half shift end to mid - 1 else start to mid + 1
                if(nums[start] <= target && target < nums[mid])
                    end = mid - 1;
                else
                    start = mid + 1;
            }
            else{
                if(nums[mid] < target && target <= nums[end])
                    start = mid + 1;
                else
                    end = mid - 1;
            }
        }
        
        return -1;
    }
};

File: Rotated_array.py
#Time Complexity : O(logn)
#Space Complexity : O(1)
#Did this code successfully run on Leetcode : YES
#Any problem you faced while coding this : NO
class Solution:
    def search(self, arr: List[int], target: int) -> int:
        low = 0
        high = len(arr) - 1
        while low <= high:
            mid = low + (high - low) //2
            if arr[mid] == target:
                return mid
            
            if arr[low] <= arr[mid]:
                if arr[low] <= target and arr[mid] >= target:
                    high = mid - 1
                else:
                    low = mid + 1
            
            else:
                if arr[high] >= target and arr[mid] <= target:
                    low = mid + 1
                else:
                    high = mid - 1
            
        return -1